sdk/go/x/auto/local_workspace.go

```diff
@@ -0,0 +1,719 @@
+) (string, string, int, error) {
+	var env []string
+	if l.PulumiHome() != nil {
+		homeEnv := fmt.Sprintf("%s=%s", pulumiHomeEnv, *l.PulumiHome())
+		env = append(env, homeEnv)
+	}
+	return runPulumiCommandSync(ctx, l.WorkDir(), env, args...)
+}
+
+// NewLocalWorkspace creates an configures a LocalWorkspace. LocalWorkspaceOptions can be used to
+// configure things like the working directory, the program to execute, and to seed the directory with source code
+// from a git repository.
+func NewLocalWorkspace(ctx context.Context, opts ...LocalWorkspaceOption) (Workspace, error) {
+	lwOpts := &localWorkspaceOptions{}
+	// for merging options, last specified value wins
+	for _, opt := range opts {
+		opt.applyLocalWorkspaceOption(lwOpts)
+	}
+
+	var workDir string
+
+	if lwOpts.WorkDir != "" {
+		workDir = lwOpts.WorkDir
+	} else {
+		dir, err := ioutil.TempDir("", "pulumi_auto")
+		if err != nil {
+			return nil, errors.Wrap(err, "unable to create tmp directory for workspace")
+		}
+		workDir = dir
+	}
+
+	if lwOpts.Repo != nil {
+		// now do the git clone
+		projDir, err := setupGitRepo(ctx, workDir, lwOpts.Repo)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create workspace, unable to enlist in git repo")
+		}
+		workDir = projDir
+	}
+
+	var program pulumi.RunFunc
+	if lwOpts.Program != nil {
+		program = lwOpts.Program
+	}
+
+	var pulumiHome *string
+	if lwOpts.PulumiHome != nil {
+		pulumiHome = lwOpts.PulumiHome
+	}
+
+	l := &LocalWorkspace{
+		workDir:    workDir,
+		program:    program,
+		pulumiHome: pulumiHome,
+	}
+
+	if lwOpts.Project != nil {
+		err := l.WriteProjectSettings(ctx, lwOpts.Project)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create workspace, unable to save project settings")
+		}
+	}
+
+	for fqsn := range lwOpts.Stacks {
+		s := lwOpts.Stacks[fqsn]
+		err := l.WriteStackSettings(ctx, fqsn, &s)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create workspace")
+		}
+	}
+
+	return l, nil
+}
+
+type localWorkspaceOptions struct {
+	// WorkDir is the directory to execute commands from and store state.
+	// Defaults to a tmp dir.
+	WorkDir string
+	// Program is the Pulumi Program to execute. If none is supplied,
+	// the program identified in $WORKDIR/pulumi.yaml will be used instead.
+	Program pulumi.RunFunc
+	// PulumiHome overrides the metadata directory for pulumi commands
+	PulumiHome *string
+	// Project is the project settings for the workspace
+	Project *workspace.Project
+	// Stacks is a map of [fqsn -> stack settings objects] to seed the workspace
+	Stacks map[string]workspace.ProjectStack
+	// Repo is a git repo with a Pulumi Project to clone into the WorkDir.
+	Repo *GitRepo
+}
+
+// LocalWorkspaceOption is used to customize and configure a LocalWorkspace at initialization time.
+// See Workdir, Program, PulumiHome, Project, Stacks, and Repo for concrete options.
+type LocalWorkspaceOption interface {
+	applyLocalWorkspaceOption(*localWorkspaceOptions)
```

Comment:
Might as well export this method and the `localWorkspaceOptions` type s.t. third parties can author their own options

Reply:
Do you mean author their own version of `NewStackLocalSource` or equivalent that takes custom options?